packages/flutter/lib/src/material/input_decorator.dart

```diff
@@ -2,22 +2,1623 @@
+///
+/// [InputDecorator] displays the visual elements of a Material Design text
+/// field around its input [child]. The visual elements themselves are defined
+/// by an [InputDecoration] object and their layout and appearance depend
+/// on the `baseStyle`, `textAlign`, `isFocused`, and `isEmpty` parameters.
+///
+/// [TextField] uses this widget to decorate its [EditableText] child.
+///
+/// [InputDecorator] can be used to create widgets that look and behave like a
+/// [TextField] but support other kinds of input.
+///
+/// Requires one of its ancestors to be a [Material] widget.
+///
+/// See also:
+///
+///  * [TextField], which uses an [InputDecorator] to display a border,
+///    labels, and icons, around its [EditableText] child.
+///  * [Decoration] and [DecoratedBox], for drawing arbitrary decorations
+///    around other widgets.
+class InputDecorator extends StatefulWidget {
+  /// Creates a widget that displays a border, labels, and icons,
+  /// for a [TextField].
+  ///
+  /// The [isFocused] and [isEmpty] arguments must not be null.
+  const InputDecorator({
+    Key key,
+    @required this.decoration,
+    this.baseStyle,
+    this.textAlign,
+    this.isFocused: false,
+    this.isEmpty: false,
+    this.child,
+  }) : assert(isFocused != null),
+       assert(isEmpty != null),
+       super(key: key);
+
+  /// The text and styles to use when decorating the child.
+  final InputDecoration decoration;
+
+  /// The style on which to base the label, hint, counter, and error styles
+  /// if the [decoration] does not provide explicit styles.
+  ///
+  /// If null, defaults to a text style from the current [Theme].
+  final TextStyle baseStyle;
+
+  /// How the text in the decoration should be aligned horizontally.
+  final TextAlign textAlign;
+
+  /// Whether the input field has focus.
+  ///
+  /// Determines the position of the label text and the color and weight
+  /// of the border.
+  ///
+  /// Defaults to false.
+  final bool isFocused;
+
+  /// Whether the input field is empty.
+  ///
+  /// Determines the position of the label text and whether to display the hint
+  /// text.
+  ///
+  /// Defaults to false.
+  final bool isEmpty;
+
+  /// The widget below this widget in the tree.
+  ///
+  /// Typically an [EditableText], [DropdownButton], or [InkWell].
+  final Widget child;
+
+  bool get _labelIsFloating => !isEmpty || isFocused;
+
+  @override
+  _InputDecoratorState createState() => new _InputDecoratorState();
+
+  /// The RenderBox that defines this decorator's "container". That's the
+  /// area which is filled if [InputDecoration.isFilled] is true. It's the area
+  /// adjacent to [InputDecoration.icon] and above the widgets that contain
+  /// [InputDecoration.helperText], [InputDecoration.errorText], and
+  /// [InputDecoration.counterText].
+  ///
+  /// [TextField] renders ink splashes within the container.
+  static RenderBox containerOf(BuildContext context) {
```

Comment:
this only seems to be used in tests?

Reply:
It was also used by the ripple (splash) code, since the ripple is clipped to the container's boundary. I took the ripple changes out of this PR to keep it smaller; I'm going to try and land the splash=>ripple changes next.